File: 1263-number-of-dice-rolls-with-target-sum/number-of-dice-rolls-with-target-sum.cpp
class Solution {
public:
    int solve(int n, int k, int target, vector<vector<long long int>> & dp){
        //base case 
        if(n<0 || target<0){
            return 0;
        }

        if( n== 0 && target == 0){
            return 1;
        }

        if(dp[n][target] != -1){
            return dp[n][target];
        }

        long long int mod = 1000000007;
        long long int ans = 0;
        for(int i=1 ; i<=k ; i++){
            ans = (ans+solve(n-1,k,target-i,dp))%mod;

        }
        return dp[n][target] =ans;
    }
    int numRollsToTarget(int n, int k, int target) {

        vector<vector<long long int>> dp(n+1,vector<long long int>(target+1,-1));
        int ans = solve(n,k,target,dp);
        return ans;

    }
};

File: 1263-number-of-dice-rolls-with-target-sum/number-of-dice-rolls-with-target-sum.java
class Solution {
    public int numRollsToTarget(int n, int k, int target) {
        long mod = 1000000007 ;
        long [][] dp = new long [n+1][target+1];

        for(long [] row : dp ){
            Arrays.fill(row,0);
        }

        dp[0][0] = 1;

        // if(dp[n][target] != 0 ){
        //     return dp[n][target];
        // }

        for(int i=1 ; i<=n ;i++){
            for(int j=1 ; j<=target ; j++){
                for( int val =1; val<=k ; val++){
                    if(j-val >= 0){
                        dp[i][j] = (dp[i][j] + dp[i-1][j-val]) % mod;

                    }
                }
            }
        }

        return (int)dp[n][target];

    }
}